### PR TITLE
[cc2538] interrupt driven radio driver

### PR DESCRIPTION
**Pull request by sjlongland**
So, I have a confession to make… in PR#3609 I accidentally included some code that would have enabled the receive interrupt on the CC2538 in the NVIC.

Given the receive interrupt handler just blindly clears the interrupt flags, this is probably not a good plan.  So, a better option, implement the rest of it, with a compile-time switch to turn interrupt driven reception off if desired.  This allows for up to two frames to be stored until such time as they can be processed by OpenThread (one frame in the radio's buffer, and one in `sReceivedFrame`).

I'm doing some more thorough tests with the GRL test suite to ensure it hasn't broken anything but indications are it is working as expected (I'd be expecting failures left-right and centre if receive was broken by this).

### CODE REVIEW

**Comment by jwhui** [comment]
@sjlongland, thank you for this contribution and the additional testing! Let us know when you've completed your testing and think this PR can move forward.

**Comment by sjlongland** [comment]
Two mistakes found:

1. I got the wrong register for the interrupt vector in the NVIC… so the interrupt service routine wasn't firing (confusingly, TI have _two_ interrupts mapped for RX/TX, and I'm not sure which one is being used)
2. In the interrupt service routine, I'm masking the constant `RFCORE_SFR_RFIRQF0` with another constant `RFCORE_SFR_RFIRQF0_RXPKTDONE` which basically means it's a no-op anyway.  `RFCORE_SFR_RFIRQF0` needs a `HWREG` macro wrapped around it.

I'll fix both of these problems. :-)

**Comment by sjlongland** [comment]
Gah… as Maxwell Smart would put it… missed it by `→ ←` that much.  `make pretty` correction pushed.

**Comment by sjlongland** [comment]
cc2538-rx-int-processing

If I surround parts of the radio driver, and comment out the call to `readFrame` from `otPlatRadioProcess` so that the only received traffic is via the interrupt, I get the above traces on a logic analyser.  In the above case:

* `PB7` was turned on just before [line 829](https://github.com/openthread/openthread/blob/4794c3fc3b23b43930a41ee89716daea841c7697/examples/platforms/cc2538/radio.c#L829) and off after [line 844](https://github.com/openthread/openthread/blob/4794c3fc3b23b43930a41ee89716daea841c7697/examples/platforms/cc2538/radio.c#L844).
* `PB2` was turned on before [line 740](https://github.com/openthread/openthread/blob/4794c3fc3b23b43930a41ee89716daea841c7697/examples/platforms/cc2538/radio.c#L740) and off after [line 824](https://github.com/openthread/openthread/blob/4794c3fc3b23b43930a41ee89716daea841c7697/examples/platforms/cc2538/radio.c#L824)
* `PB6` was turned on at [line 760](https://github.com/openthread/openthread/blob/4794c3fc3b23b43930a41ee89716daea841c7697/examples/platforms/cc2538/radio.c#L760) and off at [line 784](https://github.com/openthread/openthread/blob/4794c3fc3b23b43930a41ee89716daea841c7697/examples/platforms/cc2538/radio.c#L784)
* `PA7` was turned on at [line 786](https://github.com/openthread/openthread/blob/4794c3fc3b23b43930a41ee89716daea841c7697/examples/platforms/cc2538/radio.c#L786) and off after [line 817](https://github.com/openthread/openthread/blob/4794c3fc3b23b43930a41ee89716daea841c7697/examples/platforms/cc2538/radio.c#L817)

**Comment by jwhui** [comment]
@sjlongland, thanks for the updates and thorough testing. Is this PR ready to move forward?

**Comment by sjlongland** [comment]
I think it should be safe… receiving frames seems to be working just fine with it turned on and it seems to have a big (positive) impact on some test cases in the GRL test harness.  If we find issues, we can always turn the feature off in the OpenThread config.

File: examples/platforms/cc2538/radio.c [hunk 1]
```diff
@@ -649,12 +661,21 @@ void otPlatRadioSetPromiscuous(otInstance *aInstance, bool aEnable)
...
     uint8_t crcCorr;
     int     i;
 
+    if (sReceiveFrame.mLength != 0)
+    {
+        /*
+         * There is already a frame present in the buffer, return early so
+         * we do not overwrite it (hopefully we'll catch it on the next run).
+         */
+        return;
```
Comment (jwhui) [inline examples/platforms/cc2538/radio.c]: style: we prefer to have a single `return` statement. Can we use `otEXPECT()` here instead?

**Pull request merged**